ledger/streaming/client/src/ledgerdb_sink.rs

```diff
@@ -0,0 +1,115 @@
+/// has already done block content, scp, and avr validation and thus is
+/// considered a trusted stream
+#[pin_project]
+pub struct BlockSink<BS: Stream<Item = StreamResult<BlockData>> + std::marker::Unpin> {
+    /// Stream of block data
+    stream: BS,
+
+    /// Message channel that sends data to the db
+    sender: mpsc::Sender<BlockData>,
+
+    /// Logger running
+    logger: Logger,
+}
+
+impl<BS: Stream<Item = StreamResult<BlockData>> + std::marker::Unpin> BlockSink<BS> {
+    pub fn new(stream: BS, logger: Logger, sender: mpsc::Sender<BlockData>) -> Self {
+        Self {
+            stream,
+            logger,
+            sender,
+        }
+    }
+
+    // Ingest blocks from stream into a sink for processing
+    async fn broadcast(&mut self) {
+        while let Some(block) = self.stream.next().await {
+            if let Ok(block) = block {
+                let index = block.block().index;
+                match self.sender.send(block).await {
+                    Ok(_) => log::debug!(self.logger, "successfully sent block {}", index),
+                    Err(_) => log::warn!(self.logger, "receiver failed to send"),
+                }
+            }
+        }
+    }
+}
+
+/// Write to the ledger in a synchronous loop that consistently polls for blocks
+pub fn sink(
+    ledger_ref: Arc<Mutex<impl Ledger>>,
+    mut receiver: mpsc::Receiver<BlockData>,
+    logger: Logger,
+) {
+    let mut ledger = ledger_ref.lock().unwrap();
+    loop {
+        if let Some(block_data) = receiver.blocking_recv() {
+            let index = block_data.block().index;
+            ledger
+                .append_block(
+                    block_data.block(),
+                    block_data.contents(),
+                    block_data.signature().clone(),
+                )
+                .unwrap();
+            log::trace!(logger, "Block {} written", index)
+        } else {
+            log::warn!(logger, "Receiver shut down!");
+            break;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mc_common::logger::{test_with_logger, Logger};
+    use mc_ledger_db::{test_utils::get_mock_ledger, Ledger};
+    use std::sync::{Arc, Mutex};
+    use tokio::runtime::Builder;
+    #[test_with_logger]
+    fn assert_blocks_arrive(logger: Logger) {
+        let source_ledger = get_mock_ledger(200);
+        let mut test_block_data: Vec<StreamResult<BlockData>> = Vec::new();
+        let (tx, rcv) = mpsc::channel(10000);
+
+        for i in 1..200 {
+            test_block_data.push(Ok(source_ledger.get_block_data(i).unwrap()));
+        }
+        let stream = futures::stream::iter(test_block_data);
+        let mut bs = BlockSink::new(stream, logger.clone(), tx);
+
+        let dest_ledger = Arc::new(Mutex::new(get_mock_ledger(1)));
+        let ledger_ref = dest_ledger.clone();
+        std::thread::spawn(move || sink(ledger_ref, rcv, logger.clone()));
```

Comment:
How about we have BlockSink own spawning a thread calling `sink()`, and the associated `mpsc::channel`? It would then track a handle to that thread, cleaning it up in `stop()`, and call `stop()` on `drop()`.

Reply:
This was a detail I'm sort of wondering about in the entire chain. As James points out the sink should be a stream chain, so I'm guessing that this is the final stream component sinking into the a final component like BlockStreamService that owns the spawning thread (and turns the crank) & manages the sink thread and the BlockSink perhaps just does the mpsc sending. 

Also, I think we all need to mind meld and think about what components "client" has. I think it's probably your GrpcSource component (remote stream origin) -> validation -> sink -> BlockStreamClient (the crank turner).

What are your thoughts?

Reply:
Yeah, let's discuss tomorrow. In the meantime, I'm thinking of this thread as an implementation detail of using LMDB with async Rust, so the thread management is the responsibility of the BlockSink, not its callers, and I'd argue they shouldn't know about the thread. This also simplifies the code, as we can then trust the sender side to be our sink, and not some random code.

Reply:
While it does seem seem natural to have the sink be the final component, there probably needs to be a stream driver at the end mainly because there's a good amount of diversity of node types in the network which would lead to a wide diversity in configuration & synchronization needs

#### Examples
The following have very different ways in which they sync their ledgers with the existing services: 
* full_service: https://github.com/mobilecoinofficial/full-service/blob/main/full-service/src/validator_ledger_sync.rs 
vs: 
* consensus: https://github.com/mobilecoinfoundation/mobilecoin/blob/master/consensus/service/src/byzantine_ledger/worker.rs

What I think could probably happen is that the blocksink could have an option to spin up a sinking thread which an end node could easily spin up in their configuration or choose the option to just consume the stream that it can decide how to execute.

Right now this is setup to only target the ledger trait and not specifically ledger db (so that the underlying implementation can be different) so it's general enough to be used by nodes not using ledgedb.